File: pkg/operator/aws/s3.go
/*
Copyright 2019 Cortex Labs, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package aws

import (
	"bytes"
	"encoding/json"
	"io/ioutil"
	"path/filepath"
	"strings"

	"github.com/cortexlabs/cortex/pkg/lib/msgpack"

	"github.com/aws/aws-sdk-go/aws"
	"github.com/aws/aws-sdk-go/service/s3"

	s "github.com/cortexlabs/cortex/pkg/api/strings"
	libs3 "github.com/cortexlabs/cortex/pkg/lib/aws/s3"
	"github.com/cortexlabs/cortex/pkg/lib/errors"
	libjson "github.com/cortexlabs/cortex/pkg/lib/json"
	"github.com/cortexlabs/cortex/pkg/lib/parallel"
	cc "github.com/cortexlabs/cortex/pkg/operator/cortexconfig"
)

func S3Path(key string) string {
	return "s3://" + filepath.Join(cc.Bucket, key)
}

func IsS3File(key string) (bool, error) {
	return IsS3FileExternal(key, cc.Bucket)
}

func IsS3Dir(dirPath string) (bool, error) {
	prefix := s.EnsureSuffix(dirPath, "/")
	return IsS3Prefix(prefix)
}

func IsS3Prefix(prefix string) (bool, error) {
	return IsS3PrefixExternal(prefix, cc.Bucket)
}

func IsS3FileExternal(key string, bucket string) (bool, error) {
	_, err := s3Client.HeadObject(&s3.HeadObjectInput{
		Bucket: aws.String(bucket),
		Key:    aws.String(key),
	})

	if IsNotFoundErr(err) {
		return false, nil
	}
	if err != nil {
		return false, errors.Wrap(err, key)
	}

	return true, nil
}

func IsS3PrefixExternal(prefix string, bucket string) (bool, error) {
	out, err := s3Client.ListObjectsV2(&s3.ListObjectsV2Input{
		Bucket: aws.String(bucket),
		Prefix: aws.String(prefix),
	})

	if err != nil {
		return false, errors.Wrap(err, prefix)
	}

	hasPrefix := *out.KeyCount > 0
	return hasPrefix, nil
}

func IsS3aPrefixExternal(s3aPath string) (bool, error) {
	bucket, key, err := libs3.SplitS3aPath(s3aPath)
	if err != nil {
		return false, err
	}
	return IsS3PrefixExternal(key, bucket)
}

func UploadBytesToS3(data []byte, key string) error {
	_, err := s3Client.PutObject(&s3.PutObjectInput{
		Body:                 bytes.NewReader(data),
		Key:                  aws.String(key),
		Bucket:               aws.String(cc.Bucket),
		ACL:                  aws.String("private"),
		ContentDisposition:   aws.String("attachment"),
		ServerSideEncryption: aws.String("AES256"),
	})
	return errors.Wrap(err, key)
}

func UploadBytesesToS3(data []byte, keys ...string) error {
	fns := make([]func() error, len(keys))
	for i, key := range keys {
		key := key
		fns[i] = func() error {
			return UploadBytesToS3(data, key)
		}
	}
	return parallel.RunFirstErr(fns...)
}

func UploadFileToS3(filePath string, key string) error {
	data, err := ioutil.ReadFile(filePath)
	if err != nil {
		return errors.Wrap(err, ErrorReadFile(filePath).Error())
	}
	return UploadBytesToS3(data, key)
}

func UploadBufferToS3(buffer *bytes.Buffer, key string) error {
	return UploadBytesToS3(buffer.Bytes(), key)
}

func UploadStringToS3(str string, key string) error {
	str = strings.TrimSpace(str)
	return UploadBytesToS3([]byte(str), key)
}

func UploadJSONToS3(obj interface{}, key string) error {
	jsonBytes, err := libjson.MarshalJSON(obj)
	if err != nil {
		return err
	}
	return UploadBytesToS3(jsonBytes, key)
}

func ReadJSONFromS3(objPtr interface{}, key string) error {
	jsonBytes, err := ReadBytesFromS3(key)
	if err != nil {
		return err
	}
	return errors.Wrap(json.Unmarshal(jsonBytes, objPtr), key)
}

func UploadMsgpackToS3(obj interface{}, key string) error {
	msgpackBytes, err := msgpack.Marshal(obj)
	if err != nil {
		return err
	}
	return UploadBytesToS3(msgpackBytes, key)
}

func ReadMsgpackFromS3(objPtr interface{}, key string) error {
	msgpackBytes, err := ReadBytesFromS3(key)
	if err != nil {
		return err
	}
	return errors.Wrap(msgpack.Unmarshal(msgpackBytes, objPtr), key)
}

func ReadStringFromS3(key string) (string, error) {
	response, err := s3Client.GetObject(&s3.GetObjectInput{
		Key:    aws.String(key),
		Bucket: aws.String(cc.Bucket),
	})

	if err != nil {
		return "", errors.Wrap(err, key)
	}

	buf := new(bytes.Buffer)
	buf.ReadFrom(response.Body)
	return buf.String(), nil
}

func ReadBytesFromS3(key string) ([]byte, error) {
	response, err := s3Client.GetObject(&s3.GetObjectInput{
		Key:    aws.String(key),
		Bucket: aws.String(cc.Bucket),
	})

	if err != nil {
		return nil, errors.Wrap(err, key)
	}

	buf := new(bytes.Buffer)
	buf.ReadFrom(response.Body)
	return buf.Bytes(), nil
}

func DeleteFromS3ByPrefix(prefix string, continueIfFailure bool) error {
	listObjectsInput := &s3.ListObjectsV2Input{
		Bucket:  aws.String(cc.Bucket),
		Prefix:  aws.String(prefix),
		MaxKeys: aws.Int64(1000),
	}

	var subErr error

	err := s3Client.ListObjectsV2Pages(listObjectsInput,
		func(listObjectsOutput *s3.ListObjectsV2Output, lastPage bool) bool {
			deleteObjects := make([]*s3.ObjectIdentifier, len(listObjectsOutput.Contents))
			for i, object := range listObjectsOutput.Contents {
				deleteObjects[i] = &s3.ObjectIdentifier{Key: object.Key}
			}
			deleteObjectsInput := &s3.DeleteObjectsInput{
				Bucket: aws.String(cc.Bucket),
				Delete: &s3.Delete{
					Objects: deleteObjects,
					Quiet:   aws.Bool(true),
				},
			}
			_, newSubErr := s3Client.DeleteObjects(deleteObjectsInput)
			if newSubErr != nil {
				subErr = newSubErr
				if !continueIfFailure {
					return false
				}
			}
			return true
		})

	if subErr != nil {
		return errors.Wrap(subErr, prefix)
	}
	return errors.Wrap(err, prefix)
}

File: pkg/workloads/lib/storage/s3.py
# Copyright 2019 Cortex Labs, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import sys
import os
import boto3
import botocore
import pickle
import json
import msgpack

from lib import util
from lib.exceptions import CortexException


class S3(object):
    def __init__(self, bucket=None, region=None, client_config={}):
        self.bucket = bucket
        self.region = region

        default_config = {
            "use_ssl": True,
            "verify": True,
            "region_name": None,
            "aws_access_key_id": None,
            "aws_secret_access_key": None,
            "aws_session_token": None,
        }

        if client_config is None:
            client_config = {}

        if region is not None:
            client_config["region_name"] = region

        merged_client_config = util.merge_dicts_in_place_no_overwrite(client_config, default_config)

        self.s3 = boto3.client("s3", **client_config)

    @staticmethod
    def deconstruct_s3_path(s3_path):
        path = util.remove_prefix_if_present(s3_path, "s3://")
        bucket = path.split("/")[0]
        key = os.path.join(*path.split("/")[1:])
        return (bucket, key)

    def hadoop_path(self, key):
        return os.path.join("s3a://", self.bucket, key)

    def blob_path(self, key):
        return os.path.join("s3://", self.bucket, key)

    def _get_dir(self, prefix, local_dir):
        prefix = util.add_suffix_unless_present(prefix, "/")
        util.mkdir_p(local_dir)
        for key in self._get_matching_s3_keys_generator(prefix):
            rel_path = util.remove_prefix_if_present(key, prefix)
            local_dest_path = os.path.join(local_dir, rel_path)
            self.download_file(key, local_dest_path)

    def _file_exists(self, key):
        try:
            self.s3.head_object(Bucket=self.bucket, Key=key)
            return True
        except botocore.exceptions.ClientError as e:
            if e.response["Error"]["Code"] == "404":
                return False
            else:
                raise

    def _is_s3_prefix(self, prefix):
        response = self.s3.list_objects_v2(Bucket=self.bucket, Prefix=prefix)
        return response["KeyCount"] > 0

    def _is_s3_dir(self, dir_path):
        prefix = util.add_suffix_unless_present(dir_path, "/")
        return self._is_s3_prefix(prefix)

    def _get_matching_s3_objects_generator(self, prefix="", suffix=""):
        kwargs = {"Bucket": self.bucket, "Prefix": prefix}

        while True:
            resp = self.s3.list_objects_v2(**kwargs)
            try:
                contents = resp["Contents"]
            except KeyError:
                return

            for obj in contents:
                key = obj["Key"]
                if key.startswith(prefix) and key.endswith(suffix):
                    yield obj

            try:
                kwargs["ContinuationToken"] = resp["NextContinuationToken"]
            except KeyError:
                break

    def _get_matching_s3_keys_generator(self, prefix="", suffix=""):
        for obj in self._get_matching_s3_objects_generator(prefix, suffix):
            yield obj["Key"]

    def _upload_string_to_s3(self, string, key):
        self.s3.put_object(Bucket=self.bucket, Key=key, Body=string)

    def _read_bytes_from_s3(self, key, allow_missing=False):
        try:
            byte_array = self.s3.get_object(Bucket=self.bucket, Key=key)["Body"].read()
        except self.s3.exceptions.NoSuchKey as e:
            if allow_missing:
                return None
            raise CortexException("bucket " + self.bucket, "key " + key) from e

        return byte_array.strip()

    def search(self, prefix="", suffix=""):
        return list(self._get_matching_s3_keys_generator(prefix, suffix))

    def put_json(self, obj, key):
        self._upload_string_to_s3(json.dumps(obj), key)

    def get_json(self, key, allow_missing=False):
        obj = self._read_bytes_from_s3(key, allow_missing)
        if obj is None:
            return None
        return json.loads(obj.decode("utf-8"))

    def put_msgpack(self, obj, key):
        self._upload_string_to_s3(msgpack.dumps(obj), key)

    def get_msgpack(self, key, allow_missing=False):
        obj = self._read_bytes_from_s3(key, allow_missing)
        if obj == None:
            return None
        return msgpack.load(obj, raw=False)

    def put_pyobj(self, obj, key):
        self._upload_string_to_s3(pickle.dumps(obj), key)

    def get_pyobj(self, key, allow_missing=False):
        obj = self._read_bytes_from_s3(key, allow_missing)
        if obj is None:
            return None
        return pickle.loads(obj)

    def upload_file(self, local_path, key):
        self.s3.upload_file(local_path, self.bucket, key)

    def download_file(self, key, local_path):
        try:
            util.mkdir_p(os.path.dirname(local_path))
            self.s3.download_file(self.bucket, key, local_path)
            return local_path
        except Exception as e:
            raise CortexException("bucket " + self.bucket, "key " + key) from e

    def zip_and_upload(self, local_path, key):
        util.zip_dir(local_path, "temp.zip")
        self.s3.upload_file("temp.zip", self.bucket, key)
        util.rm_file("temp.zip")

    def download_and_unzip(self, key, local_dir):
        util.mkdir_p(local_dir)
        local_zip = os.path.join(local_dir, "zip.zip")
        self.download_file(key, local_zip)
        util.extract_zip(local_zip, delete_zip_file=True)
